redis/src/aio/async_transactions.rs

```diff
@@ -0,0 +1,127 @@
+use std::pin::Pin;
+
+use crate::{aio::ConnectionManager, cmd, pipe, Pipeline, RedisError, ToRedisArgs};
+use futures::Future;
+
+/// This function encapsulates the boilerplate required to establish a Redis transaction.
+/// Do not use it directly but use the `transaction_async!` macro instead.
+/// See the `transaction_async!` macro for more details.
+pub async fn transaction_async<
+    T,
+    E: From<RedisError>,
+    K: ToRedisArgs,
+    F: FnMut(
+        ConnectionManager,
+        &mut Pipeline,
+    ) -> Pin<Box<dyn Future<Output = Result<Option<T>, E>> + Send>>,
+>(
+    mut mgr: ConnectionManager,
+    keys: &[K],
+    mut func: F,
+) -> Result<T, E> {
+    loop {
+        cmd("WATCH").arg(keys).query_async(&mut mgr).await?;
+        let mut p = pipe();
+        let response: Option<T> = func(mgr.clone(), p.atomic()).await?;
+        match response {
+            None => {
+                println!("retrying transaction");
+                continue;
+            }
+            Some(response) => {
+                // make sure no watch is left in the connection, even if
+                // someone forgot to use the pipeline.
+                cmd("UNWATCH").query_async(&mut mgr).await?;
+                return Ok(response);
+            }
+        }
+    }
+}
+
+/// Asynchronous transaction macro for Redis operations.
+///
+/// This macro encapsulates the boilerplate required to establish a Redis transaction
+/// and apply a function to it. What it
+/// does is automatically watching keys and then going into a transaction
+/// loop util it succeeds.  Once it goes through the results are
+/// returned.
+///
+/// To use the transaction two pieces of information are needed: a list
+/// of all the keys that need to be watched for modifications and a
+/// closure with the code that should be execute in the context of the
+/// transaction.  The closure is invoked with a fresh pipeline in atomic
+/// mode.  To use the transaction the function needs to return the result
+/// from querying the pipeline with the connection.
+///
+/// The end result of the transaction is then available as the return
+/// value from the function call.
+///
+/// # Parameters
+///
+/// - `$mgr`: A cloned connection manager for Redis.
+/// - `$key`: A key (or array of keys) for the Redis operation.
+/// - `$func`: Either a path to a function or a lambda function. This function
```

Comment:
clarify that this function might be executed more than once, and that should be taken into account if it contains side effects.